File: C/Stack/LinkedList_Stack/STACK.c
#include "STACK.h"

void STACK_init(STACK *stack)
{
    stack->size = 0;
    stack->top = NULL;
}

int STACK_Size(STACK *stack)
{
    return stack->size;
}

void push(STACK *stack, int value)
{
    Node *newNode = createNode(value);
    newNode->next = stack->top;
    stack->top = newNode;
    stack->size++;
}

int pop(STACK *stack)
{
    if (stack->top == NULL)
    {
        return -1;
    }
    Node *toDelete = stack->top;
    int val = toDelete->data;
    stack->top = toDelete->next;
    free(toDelete);
    stack->size--;
    return val;
}

int top(STACK *stack)
{
    return stack->top->data;
}

int isEmpty(STACK *stack)
{
    return stack->size == 0;
}

File: Java/Stack/Linked_List_Stack/Stack.java
package Java.Stack.Linked_List_Stack;

public class Stack {

    private Node head;
    private int length;

    private class Node {
        int data;
        Node next;

        Node(int data) {
            this.data = data;
            this.next = null;
        }
    }

    Stack() {
        this.head = null;
        this.length = 0;
    }

    Stack(int value) {
        this.head = new Node(value);
        this.length = 1;
    }

    void push(int value) {
        Node newNode = new Node(value);
        if (this.empty()) {
            this.head = newNode;
        } else {
            newNode.next = this.head;
            this.head = newNode;
        }
        this.length++;
    }

    void pop() {
        if (this.empty()) {
            return;
        } else if (this.length == 1) {
            this.head = null;
        } else {
            this.head = this.head.next;
        }
        this.length--;
    }

    boolean empty() {
        return this.length == 0;
    }

    int size() {
        return this.length;
    }

    int top() {
        return this.head.data;
    }
}
